model/interactive/src/main/proto/beam_interactive_api.proto

```diff
@@ -0,0 +1,159 @@
+  // starts the stream if it is not RUNNING then pauses the stream when the
+  // offset is reached.
+  rpc Advance (AdvanceRequest) returns (AdvanceResponse) {}
+
+  // Stops and resets the stream to the beginning.
+  rpc Stop (StopRequest) returns (StopResponse) {}
+
+  // Pauses the stream of events to the EventsRequest. If there is already an
+  // outstanding EventsRequest streaming events, then the stream will pause
+  // after the EventsResponse is completed.
+  // To un-pause, send either a Start or Advance request.
+  rpc Pause (PauseRequest) returns (PauseResponse) {}
+
+  // Sends a single element to the EventsRequest then closes the stream.
+  rpc Step (StepRequest) returns (StepResponse) {}
+
+  // Responds with debugging and other cache-specific metadata.
+  rpc Status (StatusRequest) returns (StatusResponse) {}
+}
+
+message StartRequest {
+  // (Optional) How quickly the stream will be played back, e.g. if
+  // playback_speed == 2, then the stream will replay events twice as fast as
+  // they were recorded. If unspecified, this will default to 1.
+  double playback_speed = 1;
+}
+message StartResponse { }
+
+message AdvanceRequest {
+  // (Required) Will advance the stream by replaying events as quickly as
+  // possible until the stream timestamp has advanced by the specified amount.
+  google.protobuf.Duration advance_by = 1;
+}
+message AdvanceResponse {}
+
+message StopRequest { }
+message StopResponse { }
+
+message PauseRequest { }
+message PauseResponse { }
+
+message StatusRequest { }
+message StatusResponse {
+  // The current timestamp of the replay stream. Is MIN_TIMESTAMP when state
+  // is STOPPED.
+  google.protobuf.Timestamp stream_time = 1;
+
+  // The minimum watermark across all of the faked replayable unbounded sources.
+  // Is MIN_TIMESTAMP when state is STOPPED.
+  google.protobuf.Timestamp watermark = 2;
+
+  // The latest timestamp of the recording stream. Is MIN_TIMESTAMP if there is
+  // no recording.
+  google.protobuf.Timestamp recording_time = 3;
+
+  // The set playback_speed from the StartRequest. Playback speed is set by
+  // StartRequest, or if the stream_time is the current time and the recording
+  // is still happening, the playback speed is 1, else 0.
+  double playback_speed = 4;
+
+  enum State {
+    // The InteractiveService is not replaying. Goes to RUNNING with a
+    // StartRequest.
+    STOPPED = 0;
+
+    // The InteractiveService is replaying events. Goes to PAUSED with a
+    // PauseRequest. Goes to STOPPED with a StopRequest.
+    RUNNING = 1;
+
+    // The InteractiveService is paused from replaying events. Goes to RUNNING
+    // with either a StartRequest or a StepRequest. Goes to STOPPED with a
+    // StopRequest.
+    PAUSED = 2;
+  }
+  State state = 5;
+}
+
+message StepRequest {
+  // (Optional) The count of events to stream. If unspecified, the default is
+  // 1, i.e. only read a single element from each source and send it through
```

Comment:
Should we read a single element from any source, instead of one element from each source? It can be impossible to read one single element from each source since you might need to read many elements from one source before an element is available from another source.

Reply:
Ack, changed to any source.